Add Amount type

stevenroose opened the pull request:
(Retake of #192 with a simpler structure for simple satoshi-precision amounts.)

### Rationale

There are a lot of places and ways in which Bitcoin amounts are used: displaying, converting denominations, arithmetic, de/serializing in different ways. It's not always clear what type to use for a Bitcoin amount: `i64`? `u64`? `f64`? All three of those are common choices and are confusing when used together.

The `Amount` type has the following functions:
functions
- safe arithmetic: `checked_*` methods for checked arithmetic and `std::ops` trait implementations for common operations that panic on over/underflow
- ability to display in all common denominations: BTC, mBTC, uBTC, bits, satoshi and msat
- instantiation from satoshi (native underlying type) or any of the above named denominations
- `serde` serialization as `i64` satoshi or as `f64` bitcoin (extensible with other options, f.e. as a string)
- parse amount strings with above mentioned denominations

The underlying type is `i64`. There was a hard discussion about whether to use `i64` or `u64`, i.e. whether or not to allow negative amounts. One of the main deciding factors was the fact that the Bitcoin Core API has several wallet-related calls where negative amounts are used. Several utility functions are added to aid those that want to avoid negative amounts: `abs`, `is_negative` and `positive_sub` (functions like `checked_sub` as if the underlying type was `u64`, i.e. returns `None` if the result or either of the operands of the subtraction would be negative).

Superseded by https://github.com/rust-bitcoin/rust-bitcoin/pull/270

stevenroose commented:
Rebased this on top of https://github.com/rust-bitcoin/rust-bitcoin/pull/250 so I could use `serde_json` in tests.

sgeisler commented on src/util/amount.rs:
```diff
@@ -0,0 +1,789 @@
+
+        let mut decimals = None;
+        let mut value: i64 = 0; // as satoshis
+        for c in s.as_bytes() {
```
Is there a good reason to use bytes instead of chars?

stevenroose replied:
Good question. Probably not. What's the exact difference? I think chars is probably safer, even though `String`s are guaranteed to be UTF8, right?

stevenroose replied:
using chars now

sgeisler commented on src/util/amount.rs:
```diff
@@ -0,0 +1,789 @@
+    ///
+    /// Note: This only parses the value string.  If you want to parse a value with denomination,
+    /// use `FromStr`.
+    pub fn parse_denom(mut s: &str, denom: Denomination) -> Result<Amount, ParseAmountError> {
```
I also wonder if we couldn't use `f64` here if we defined the maximum valid range as `-21*10^14` sat till `21*10^14 sat`. `log(21*10^(6+8), 2) ≈ 50.9` and a `f64` has 52 mantissa bits, that should be enough to not have any losses imo (but I'd need to test this as well, but with only `2^51` combinations even exhaustive testing seems feasible).

stevenroose replied:
Hmm, I never touched the fuzz tests. I can try to look into this. Could also create random amounts (uniform i64 values) and string and destring for all denominations etc.

apoelstra replied:
`f64` in principle has enough precision to represent any bitcoin amount, yes. (This is probably one reason for Satoshi choosing the "max 21mil with 8 decimal places" limits he did.) But in practice it's hard to write a parser that doesn't do things like dividing by 10 that create rounding errors and result in wrong numbers being parsed anyway.

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,789 @@
+
+// No one should depend on a binding contract for Display for this type.
+// Just using Bitcoin denominated string.
+impl fmt::Display for Amount {
```
Display is certainly useful. FromStr, I can agree that it's doubtable. Because in all cases, you'll probably care about what format is allowed. (While with Display often times (logging, printing) you don't really care.)

stevenroose replied:
We can add any number of alternative spellings to `FromStr for Denomination`. But I know what you mean. It's not exact science. It's mostly useful to make sure you can at least roundtrip your own strings. Which I think can have some value.

stevenroose replied:
I added `sat` parsing.

stevenroose replied:
Well, to be honest, `Display` and `Debug` are meant for human readability. So all you should expect from them is that you'll be able to understand the amount from seeing it's string representation.
I don't think just a number would be sufficient for `Display`. There is `as_sat()` is you just want a satoshi number and `to_string_in` if you want just the number in any other denomination. If you want I could add a shorthand `as_btc()` that returns `as_float_denom(Denomination::Bitcoin)`, but I think most people wanting the bitcoin value is just for printing, so `to_string[_in]()` would do.

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,789 @@
+    }
+}
+
+impl ops::Mul<i64> for Amount {
```
done

jonasnick commented on src/util/amount.rs:
```diff
@@ -0,0 +1,798 @@
+
+    #[test]
+    fn into_btc() {
+        let amt: Amount = 0.25.into_btc(); // type annotaion needed
```
annotation typo

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,798 @@
+    /// Express this Amount as a floating-point value in the given denomination.
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn as_float_denom(&self, denom: Denomination) -> f64 {
```
See my comment above about a possible `as_btc()` shortcut. About you comment: I think nowadays a lot of wallets let you chose which denomination to show amounts in, so I think it has some value to be able to do this without having to write the rescaling again. I realize we should of course not encourage using floating points, but I think the warning in the documentation should be enough.

jonasnick replied:
I wasn't suggesting to get rid of denominations, just that `f64`'s should be always bitcoin for compatibility with getbalance, sendtoaddress, etc. But I see now that there's no way to get integers other than `as_sat`. Maybe `as_mbtc`, `as_bit`, etc. which return integers and error if this isn't possible would be useful.

stevenroose replied:
Hmm, what use cases do you see for converting to other denominations as integers? You're always gonna lose information except in the exception of amounts that have trailing zeroes.

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,798 @@
+
+// The inner type used to represent amounts.
+// This is done to more easily change the underlying type in case this is desired in the future.
+type Inner = i64;
```
I was thinking about this too. I'm still torn on it. There are pieces in Bitcoin Core where negative amounts are used for sure (I can tell just converted all those places to use Asset-Value maps :D). Mostly fees or credit/debit of txs for the wallet. A way in the middle I could perhaps add methods `is_positive()` or `sign()` (returning -1, 0 or 1) or `abs()`.

I'm open to opinions on this point.

Another thing I noticed is that I end up writing a lot of `Amount::from_sat(xxx as i64)` because all serialized/stored values are unsigned. I think of this more of an inconvenience to type than a possible issue with overflows, but well.

stevenroose replied:
changed to u64, resolving this

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,798 @@
+}
+
+impl ops::Add for Amount {
+    type Output = Amount;
```
Can you do that? Like `(amt1 + amt2)?`?

stevenroose replied:
I can use `CheckedAdd`. I can actually implement all the `Checked*` traits.

stevenroose replied:
Ah they don't seem to exist in stdlib. Just `checked_*` methods on primitive types.

jonasnick replied:
There is `checked_add` in `std`.

jonasnick replied:
Fwiw, didn't see your updates when I wrote above comment

stevenroose replied:
Yeah but not as a trait. So yeah we could just have those method in the `impl`, but it's not a trait.

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,798 @@
+    type Output = Amount;
+
+    fn mul(self, rhs: f64) -> Self::Output {
+        Amount::from_inner((self.0 as f64 * rhs) as Inner)
```
I added `Rem<i64>`. My main issue with the float operators is that they are impossible to do totally correct and are just there for convenience, not for any high-reliability usage. Also, considering usage of amounts between -21 million and 21 million, I think @sgeisler pointed out that floating points can correctly represent all values up to satoshi precision.

jonasnick replied:
I don't think there's a place in Bitcoin where you favor convenience over correctness. Why allow operating with `f64` at all then? If we do then we should make it as good as possible. Also, there's no way to divide without using f64 right now. We could add Div<i64> and check that remainder is 0.

stevenroose replied:
How about dropping `Mul<f64>` altogether? There is `Mul<i64>` anyway.

What do you think is the best way to do a checked div? Using `Div<i64>` and `Output = Result<Amount, Error>` is really inconvenient, IMO. I can mimick the `checked_*` methods from std.

jonasnick replied:
Oh, I guess you mean that matching `(amt * 10)` is inconvenient. I think that's fine - you do need to be careful when dealing with amounts. I'd remove the unchecked arithmetic.

stevenroose replied:
Hmm, unchecked non-float arithmetic still panics when overflowing, so as long as we protect users from silent overflows, whether they want to check or have their program panic, I think can be up to them. I'm ok with removing all float arithmetic. Of users really want it, they can get the float and do their math on it.

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,913 @@
+    /// Convert an amount in floating-point notation with a given denomination.
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn from_float_denom(value: f64, denom: Denomination) -> Amount {
```
Good catch!

stevenroose replied:
Check [my latest comment](https://github.com/rust-bitcoin/rust-bitcoin/pull/252/#issuecomment-482071581) regarding this. I solved it somehow, but I'm not very happy with it.

stevenroose commented:
Ok so I updated the PR to use `u64` instead of `i64` as discussed on IRC. I also removed float arithmetic, and changed the `std::ops` implementation to panic on overflow, which I think is a good middle ground between safety and convenience.

The one last thing I'm not fully content with is the `from_float_denom` implementation. I want to return `TooPrecise` when the amount has some significant digits, but still avoid failing to convert `1.0 + 3.0` because internally it's `4.00000000000000012` or so. I didn't feel comfortable to introduce arbitrary constants like `sats.fract() > 1000.0 * f64::EPSILON`. 
Suggestions welcome. I think it's valuable to have the method. But in principle I could just implement it as stringing the float and parsing it.

jonasnick commented on src/util/amount.rs:
```diff
@@ -0,0 +1,901 @@
+            return Err(ParseAmountError::Negative);
+        }
+        if value > Amount::max_value().as_float_denom(denom) {
+            return Err(ParseAmountError::TooBig);
```
I think there is something wrong here because the following works:
```
assert_eq!(f(18446744073709551617.0, D::Satoshi), Ok(Amount(0)));
```

stevenroose replied:
Yeah, I've noticed this. At the u64::max_value level, f64 doesn't have enough precision to distinguish a single satoshi :/

I think instead of using `u64::max_value()`, I'll look up what is the highest number f64 can represent with at least 1.0 precision. And then error if it's bigger than that.

I actually re-implemented as_btc parsing to work without floats and added some more edge-case tests. Let me port over those tests.

stevenroose replied:
I added you test case and it returns the correct error now.

jonasnick commented on src/util/amount.rs:
```diff
@@ -0,0 +1,901 @@
+
+    /// Convert from a value expressing bitcoins to an Amount.
+    pub fn from_btc<T: IntoBtc>(btc: T) -> Result<Amount, ParseAmountError> {
+        btc.into_btc()
```
I don't understand the name`from_btc`. Shouldn't this be called `from_nonbtc`?

stevenroose replied:
No, the `IntoBtc` trait means that you can convert the type interpreting the value as bitcoin into an Amount. This is taken from `rust-bitcoin-amount`, tbh. I could just remove it all.

So `from_btc(5.0)` makes a 5 bitcoin amount, so it's equivalent with `from_sat`.

The `IntoBtc` trait impl on `f64` or `serde_json::Number` mean "convert this number into Amount as if it was bitcoins". So `5.0.into_btc()` creates a 5 bitcoin Amount.

stevenroose replied:
Perhaps `AsBtc` was a better name, but again this is still there because I started off from the bitcoin-amount crate. I'm open to removing it, that would also remove our dependency on `serde_json`.
@apoelstra 
Let me just do that. I'll remove the trait and make this method take `f64`.

jonasnick replied:
Probably I wasn't expressing myself clearly. If `5.0.into_btc()` creates a 5 bitcoin amount that implies that the literal `5.0` is not `btc` because we first have to convert it into `btc`. How can we use a non-`btc` thing (`5.0`) as argument to `from_btc` whose name would imply that we give it a `btc` thing?

dpc replied:
I would like to add that `frm_btc` is totally confusing. I look at the signature, I look at the comment, and I have no idea what it is supposed to do.

stevenroose replied:
I removed the IntoBtc trait.

dpc commented on src/util/amount.rs:
```diff
@@ -0,0 +1,901 @@
+    /// Equivalent to `as_float_denom(Denomination::Bitcoin)`.
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn as_btc(&self) -> f64 {
```
Just `to_float`? The "standard" (Bitcoin) denomination is kind of implied and seems natural, and is `_btc` is just confusing.

stevenroose replied:
I don't agree with that. There is not "default" denomination and there should not be. If there was we'd have `From` and `To` and the serde serialize traits.

dpc replied:
Well, maybe but `to_btc` is super-confusing, because "btc" is so vague, and it is not immediately obvious that it means "denomination".

`to_float_in_btc`?

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,901 @@
+    ///
+    /// Note: This only parses the value string.  If you want to parse a value with denomination,
+    /// use [FromStr].
+    pub fn parse_denom(mut s: &str, denom: Denomination) -> Result<Amount, ParseAmountError> {
```
Changed.

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,901 @@
+    /// Express this Amount as a floating-point value in the given denomination.
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn as_float_denom(&self, denom: Denomination) -> f64 {
```
Not really sure about `to`/`as`, since you're not really turning this in something else, but just representing the inner value as another type. Anyway, will change. I agree with denomination naming.

stevenroose replied:
Hmm, similar to `to_string_in`, I opted for `to_float_in` and `from_str_in`. Do you agree with that?

stevenroose commented:
I made some updates:
- rebased off master instead of off https://github.com/rust-bitcoin/rust-bitcoin/pull/250 (the no-strason PR)
- removed the `IntoBtc` trait, which removes the `serde_json` non-dev dependency
- `from_btc` now just takes `f64`
- removed `denom` from all method names, the convention is now `(from|to)_*_in`
- `from_float_in`: replaced direct float interpretation with parsing the `to_string` impl of `f64`; anyone willing to come up with a sane way to do f64 -> Amount conversion is very much encouraged to contribute so

dpc commented on src/util/amount.rs:
```diff
@@ -0,0 +1,868 @@
+
+    /// Get a formatted string of this Amount in the given denomination, suffixed with the
+    /// abbreviation for the denomination.
+    pub fn to_string_with_denomination(&self, denom: Denomination) -> String {
```
Isn't this the same thing as `to_string_in` right above it?

stevenroose replied:
No, `to_string_in` is just the amount (i.e. `0.42`), this is "with denomination" (i.e. `0.42 BTC`). I think the former has value.

dpc replied:
Oh, sorry, my bad. Agreed. :)

dpc commented:
I see that you removed `IntoBtc` while I was commenting on it. Pasting it below.


Nit, but "A trait used to" is redundant. Reader sees that it is a trait. :) "Can be converted to ..." is shorter and to the point.


This "into btc" is confusing. Shouldn't this be just "TryIntoAmount"? Also - shouldn't the error be associated type? Why would it always had to be "Parse"? That would imply using for just parseable stuff like strings. 

`TryInto` etc. was just stabilized, but rust-bitcoin won't be able to use it for a while, but we could just copy that trait literally, and in the future `impl<T> TryInto<Amount> for T where T: TryIntoAmount`, and eventually deprecate `TryIntoAmount`.

stevenroose commented:
Lol meta-discussion. The `IntoBtc` name meant that the value was supposed to be interpreted as "BTC". A `TryIntoAmount` for `f64` wouldn't tell you how many satoshi's you would get if you used it for `4.2`, f.e..

But well, it's removed because it was not super valuable and caused confusion apparently.

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,863 @@
+//! In a few functions, this module supports convertion to and from floating-point numbers.
+//! Please be aware of the risks of using floating-point numbers for financial applications.
+//! These types of numbers do not give any guarantee to retain the precision
+//! of the original amount when converting, or when doing arithmetic operations.
```
Ah yeah that's a good point. Let me quickly go over all documentation again to see if something changed.

stevenroose replied:
Updated the docs. I moved the important piece to the `Amount` type instead of the module because I think most users will find this amount through a re-export as `bitcoin::Amount`.

jonasnick replied:
updated doc looks good to me

sgeisler commented on src/util/amount.rs:
```diff
@@ -0,0 +1,876 @@
+        assert!(t.unwrap_err().to_string().contains(&ParseAmountError::Negative.to_string()));
+        let t: Result<T, serde_json::Error> = serde_json::from_str("{\"amt\": 1000000.000000001}");
+        assert!(t.unwrap_err().to_string().contains(&ParseAmountError::TooPrecise.to_string()));
+        //let t: Result<T, serde_json::Error> = serde_json::from_str("{\"amt\": 21000000.000000001}");
```
Can this be removed?

stevenroose replied:
Well, it's a test that should work, but the fact that serde forces us to do floating point makes it not work.

sgeisler replied:
If you think we'll be able to use the code in the future can you make it a `FIXME` or `TODO` please?  If not removing it seems appropriate.

apoelstra replied:
I don't think we'll be able to use it in the foreseeable future. We ought to just remove it.

stevenroose replied:
Removed.

sgeisler commented on src/util/amount.rs:
```diff
@@ -0,0 +1,876 @@
+
+    /// Checked addition.
+    /// Returns [None] if overflow occurred.
+    pub fn checked_add(self, rhs: Amount) -> Option<Amount> {
```
The `checked_*` stuff is great!

stevenroose replied:
I thought it was sad there are no traits for it, only in the `num` crate. But yeah it's good. We've been using it already internally.

sgeisler replied:
We could have an optional dependency on `num-traits`, I'm using it for `rust-lightning-invoice`and think it's an acceptable dependency in many cases.

stevenroose replied:
Just out of curiosity: how would you do an optional trait impl with a method signature identical to a native method of the type? 
```rust
#[cfg(feature = "num-traits")]
impl CheckedAdd for Amount {
  fn checked_add(self, other: Amount) -> Amount {
    self.checked_add(other)
  }
}
```
This looks like an infinite loop..

sgeisler replied:
I think you can differentiate them by calling `Amount::checked_add(self, other)` or `CheckedAdd::checked_add(self, other)`.

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,876 @@
+                '.' => 1,
+                _ => 2,
+            };
+            if s.chars().skip(1).map(penalty).sum::<i32>() > 1 {
```
How about an extra sanity length check? u64::max_value is `18446744073709551615` (20 digits). I'm inclined to just error `InvalidFormat` on anything that has size >50.
But agreed it should be `usize`.

sgeisler replied:
Yes, a length check would avoid certain DoS scenarios and would make the code fail early before iterating over all characters.

stevenroose commented on src/util/amount.rs:
```diff
@@ -0,0 +1,876 @@
+            if s.chars().skip(1).map(penalty).sum::<i32>() > 1 {
+                return Err(ParseAmountError::InvalidFormat);
+            }
+            return Err(ParseAmountError::Negative);
```
@apoelstra Opinion? I could just have a small check that checks if size == 0 or size > 50 and return InvalidFormat and then skip the `-` check. If then a negative number is encountered from an RPC or some other place, it can be harder to debug. Personally I think a small performance hit in the error case it not a big deal.
But I like the simpler error handling. Less error cases to check for the caller.

Like this:
```diff
diff --git a/src/util/amount.rs b/src/util/amount.rs
index ef02413..0cf1d4d 100644
--- a/src/util/amount.rs
+++ b/src/util/amount.rs
@@ -193,23 +193,10 @@ impl Amount {
     /// Note: This only parses the value string.  If you want to parse a value
     /// with denomination, use [FromStr].
     pub fn from_str_in(mut s: &str, denom: Denomination) -> Result<Amount, ParseAmountError> {
-        if s.len() == 0 {
+        if s.len() == 0 || s.len() > 50 {
             return Err(ParseAmountError::InvalidFormat);
         }
 
-        if s.chars().nth(0).unwrap() == '-' {
-            // Quickly check whether the remainder of the string is only numbers and max one point.
-            let penalty = |c| match c {
-                '0'...'9' => 0,
-                '.' => 1,
-                _ => 2,
-            };
-            if s.len() == 1 || s.chars().skip(1).map(penalty).sum::<usize>() > 1 {
-                return Err(ParseAmountError::InvalidFormat);
-            }
-            return Err(ParseAmountError::Negative);
-        }
-
         let max_decimals = {
             // The difference in precision between native (satoshi)
             // and desired denomination.
```

sgeisler replied:
Maybe there could be an `InvalidCharacter(char)` error that tells you what went wrong? That would not only help if an API returns a negative number but e.g. also if it returns a number with a comma instead of a decimal dot.

apoelstra replied:
Yeah, I'd like `InvalidCharacter(char)`. No strong opinions on also having `Negative` ... I wouldn't put the effort in to code that, but I don't mind reviewing it if somebody else does ;)

stevenroose replied:
Added `InvalidCharacter(char)`. Also removed the negative check, it was getting a bit too complex (I tried folding with an accumulator, but not really worth it).

stevenroose commented:
What else is blocking this?

sgeisler commented on src/util/amount.rs:
```diff
@@ -0,0 +1,868 @@
+    ///
+    /// Note: This only parses the value string.  If you want to parse a value
+    /// with denomination, use [FromStr].
+    pub fn from_str_in(mut s: &str, denom: Denomination) -> Result<Amount, ParseAmountError> {
```
I began verifying the parsing code by trying to build a DFA from it to derive a regular expression. The regex seems ok: `[0-9]*.[0-9]{0,precision_diff}` the only weird case this allows is parsing `.` as `0`.

But the precision verification is broken, e.g. the following test case 
```rust
assert_eq!(p("12.000", Denomination::MilliSatoshi), Err(ParseAmountError::TooPrecise));
```
fails with
```
left: `Ok(Amount(12 satoshi))`,
 right: `Err(TooPrecise)`', src/util/amount.rs:712:9
```

stevenroose replied:
That's a good catch! I fixed that in an extra commit. The string roundtrip test didn't catch this because the `to_string_in` was also broken :/

sgeisler commented on src/util/amount.rs:
```diff
@@ -0,0 +1,858 @@
+                // precise amount.  That is not allowed unless there are no decimals and
+                // the last digits are zeroes as many as the diffence in precision.
+                let last_n = precision_diff.abs() as usize;
+                if s.contains(".") || !s.chars().skip(s.len() - last_n).all(|d| d == '0') {
```
That might be personal preference, but to me
```rust
s.chars().rev().take(last_n).any(|d| d != '0')
```
looks much more straight forward than
```rust
!s.chars().skip(s.len() - last_n).all(|d| d == '0')
```

stevenroose replied:
Hmm, yeah I guess it's personal. I think the most important thing is that the comment above it explains what it is supposed to do. I think `all [...] == '0'` is more clear than `any [...] != '0'`. So how about a combination of both:
```rust
!s.chars().ref().take(last_n).all(|d| d == '0')
```

stevenroose replied:
meh, while typing I settled for you suggestion :D

apoelstra commented:
concept ACK

jonasnick commented:
@stevenroose serde tests fail according to travis

apoelstra commented on src/util/amount.rs:
```diff
@@ -0,0 +1,921 @@
+    }
+
+    /// Get the number of satoshis in this [Amount].
+    pub fn as_sat(self) -> i64 {
```
Can we have an `as_usat` or something that returns a u64? I see in @jonasnick's code using this that there are zillions of casts to u64.

stevenroose replied:
One that panics on negative or one that `Result`s? I'd personally do `try_as_usat() -> Option<u64>` and `as_usat() -> u64`.

apoelstra replied:
:( in my usage negative amounts never ever exist. It's really unergonomic to have these extra error paths and also have all my amounts potentially being negative. It also prevents computing fees using `checked_sub` to catch unbalanced transactions as a special case.

stevenroose commented:
Please see https://github.com/rust-bitcoin/rust-bitcoin/pull/270 for a replacement PR that has explicit `u64` `Amount` and `i64` `SignedAmount`.
